base16ct/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+#![no_std]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![forbid(unsafe_code)]
+#![warn(missing_docs, rust_2018_idioms)]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg",
+    html_root_url = "https://docs.rs/base16ct/0.0.0"
+)]
+#![doc = include_str!("../README.md")]
+
+//! # Implementation
+//!
+//! Implemented using integer arithmetic alone without any lookup tables or
+//! data-dependent branches, thereby providing portable "best effort"
+//! constant-time operation.
+//!
+//! Not constant-time with respect to message length (only data).
+//!
+//! Adapted from this C++ implementation:
+//!
+//! <https://github.com/Sc00bz/ConstTimeEncoding/blob/master/hex.cpp>
+//!
+//! Copyright (c) 2014 Steve "Sc00bz" Thomas (steve at tobtu dot com)
+//! Derived code is dual licensed MIT + Apache 2.0 (with permission from @Sc00bz)
+
+#[cfg(feature = "alloc")]
+#[macro_use]
+extern crate alloc;
+#[cfg(feature = "std")]
+extern crate std;
+
+use core::fmt;
+
+#[cfg(feature = "alloc")]
+use alloc::{string::String, vec::Vec};
+
+/// Error type
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Error {
+    /// Invalid encoding of provided Base16 string.
+    InvalidEncoding,
+
+    /// Insufficient output buffer length.
+    InvalidLength,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidEncoding => f.write_str("invalid Base16 encoding"),
+            Error::InvalidLength => f.write_str("invalid Base16 length"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+/// Base16 encoding (a.k.a. hexadecimal)
+#[derive(Copy, Clone, Debug, Default, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct Base16 {
+    /// Upper or lower case
+    case: Case,
+}
+
+impl Base16 {
+    /// Lower case hex: 0-9 a-f
+    pub fn lower_case() -> Base16 {
+        Base16 { case: Case::Lower }
+    }
+
+    /// Upper case hex: 0-9 A-F
+    pub fn upper_case() -> Base16 {
+        Base16 { case: Case::Upper }
+    }
+}
+
+impl Base16 {
+    /// Decode a Base16 (hex) string into the provided destination buffer.
+    pub fn decode<'a>(&self, src: impl AsRef<[u8]>, dst: &'a mut [u8]) -> Result<&'a [u8], Error> {
+        let src = src.as_ref();
+
+        let dst_length = Self::decoded_len(src)?;
+        if dst_length > dst.len() {
+            return Err(Error::InvalidLength);
+        }
+
+        let mut err: usize = 0;
+
+        for (i, dst_byte) in dst.iter_mut().enumerate().take(dst_length) {
```

Comment:
Personally I think `dst[..dst_len].iter_mut().enumerate()` is easier to read. Alternatively you could use something like this:
```rust
for (src, dst) in src.chunks_exact(2).zip(dst.iter_mut()) { .. }
```
Zipping should handle the `dst` truncation.

Reply:
Also the last suggestion helps compiler to remove unreachable panics.